foyer/forcefield.py

```diff
@@ -317,9 +328,42 @@ def apply(self, topology, references_file=None, *args, **kwargs):
             positions = np.empty(shape=(topology.getNumAtoms(), 3))
             positions[:] = np.nan
         box_vectors = topology.getPeriodicBoxVectors()
-        system = self.createSystem(topology, *args, **kwargs)
+        system, data = self.createSystem(topology, *args, **kwargs)
 
         structure = pmd.openmm.load_topology(topology=topology, system=system)
+
+        '''
+        Check that all topology objects (angles, dihedrals, and impropers)
+        have parameters assigned. OpenMM will generate an error if bond parameters
+        are not assigned.
+        '''
+        if data.angles and (len(data.angles) != len(structure.angles)):
+            msg = ("Parameters have not been assigned to all angles. Total "
+                   "system angles: {}, Parameterized angles: {}"
+                   "".format(len(data.angles), len(structure.angles)))
+            if assert_angle_params:
+                raise Exception(msg)
+            else:
+                warnings.warn(msg)
+        if data.propers and len(data.propers) != \
+                len(structure.dihedrals) + len(structure.rb_torsions):
+            msg = ("Parameters have not been assigned to all proper dihedrals. "
+                   "Total system dihedrals: {}, Parameterized dihedrals: {}"
+                   "".format(len(data.propers), len(structure.dihedrals) + \
+                   len(structure.rb_torsions)))
+            if assert_dihedral_params:
+                raise Exception(msg)
+            else:
+                warnings.warn(msg)
+        if data.impropers and (len(data.impropers) != len(structure.impropers)):
+            msg = ("Parameters have not been assigned to all impropers. Total "
+                   "system impropers: {}, Parameterized impropers: {}"
+                   "".format(len(data.impropers), len(structure.impropers)))
+            if assert_improper_params:
+                raise Exception(msg)
```

Comment:
Could you factor these 3 checks out into a consistent "error or warn" function? Lots of repeated code here

Reply:
Good call, should help to condense this